src/unistd.rs

```diff
@@ -511,6 +511,40 @@ pub fn sleep(seconds: libc::c_uint) -> c_uint {
     unsafe { libc::sleep(seconds) }
 }
 
+/// Creates a regular file which persists even after process termination
+///
+/// * `template`: a path whose 6 rightmost characters must be X, e.g. /tmp/tmpfile_XXXXXX
+/// * returns: tuple of file descriptor and filename
+///
+/// Err is returned either if no temporary filename could be created or the template doesn't
+/// end with XXXXXX
+///
+/// # Example
+///
+/// ```rust
+/// use nix::unistd;
+///
+/// let fd = match unistd::mkstemp("/tmp/tempfile_XXXXXX") {
+///     Ok((fd, path)) => {
+///         unistd::unlink(path.as_path()).unwrap(); // flag file to be deleted at app termination
+///         fd
+///     }
+///     Err(e) => panic!("mkstemp failed: {}", e)
+/// };
+/// // do something with fd
+/// ```
+#[inline]
+pub fn mkstemp<P: ?Sized + NixPath>(template: &P) -> Result<(RawFd, PathBuf)> {
+    let mut path = try!(template.with_nix_path(|path| {path.to_bytes_with_nul().to_owned()}));
+    let p = path.as_mut_ptr() as *mut _;
+    unsafe {
+        let fd = libc::mkstemp(p);
+        let pathname = OsStr::from_bytes(CStr::from_ptr(p).to_bytes());
+        try!(Errno::result(fd));
+        Ok((fd, PathBuf::from(pathname).to_owned()))
```

Comment:
`PathBuf::from` already creates an owned PathBuf, so `to_owned()` is just making an additional copy. Also, it's unnecessary since `path` is already owned in the first place; you can just strip out the trailing nul byte and use `OsString::from_vec`